### PR TITLE
Return target factor sequence scores in beam search & scoring

### PR DESCRIPTION
**Pull request by fhieber**
Adds recording of target factor scores in beam search and for scoring. Sequence scores for secondary target factors remain unnormalized and are returned separately from the primary hypothesis scores (which are normalized).
Downstream applications can decide how to use these scores.

As part of this change a few internal classes / interfaces were changed:
- search algorithms now return a dataclass `SearchResult`
- `score` attribute of `Translation` class changed to `scores` holding all target-side scores. Index 0: primary hypothesis score, Indices 1+: scores for other factor sequences
- `score` attribute of `NbestTranslations` class changed to `scores`
- `factor_scores` attribute added to `TranslatorOutput`

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by xingniu** [comment]
(1) Does it make sense to provide both word scores and TF scores (instead of summing them directly)? I'm not sure if simply summing them could make score calibration worse.
(2) TFs are generally easy to predict, so per-token TF scores in a segment could be very close. Could max or min (depending on the sign) be more informative?

**Comment by fhieber** [comment]
You have a good point about score calibration. Let me make this WIP and reconsider. Returning separate scores for each factor sequence is likely cleanest and allows users to decide how to combine the scores.

**Comment by fhieber** [comment]
@xingniu I updated the PR. Change turned out to be a bit larger than I initially expected but I agree with you that returning the scores separately is the better choice.

File: sockeye/score_pt.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def score(args: argparse.Namespace):
         logger.info("CUDA not available, using cpu")
         use_cpu = True
     device = pt.device('cpu') if use_cpu else pt.device('cuda', args.device_id)
-    logger.info("Scoring device: {device}")
+    logger.info(f"Scoring device: {device}")
```
Comment (mjdenkowski) [inline sockeye/score_pt.py:59]: Good catch!

**Review by mjdenkowski — APPROVED** [review]
Looks good!

There are only a few minor comment issues to address before merging.

**Comment by mjdenkowski** [comment]
It looks like we also need to update the version.

File: sockeye/beam_search_pt.py [hunk 2]
```diff
@@ -77,13 +77,23 @@ def decode_step(self,
...
-            # target factors are greedily 'decoded'.
-            factor_predictions = [tfo.argmax(dim=1).unsqueeze(1).int() for tfo in target_factor_outputs]
-            target_factors = factor_predictions[0] if len(factor_predictions) == 1 else pt.cat(factor_predictions, 1)
+            predictions = []  # type: List[pt.Tensor]
+            for tf_logits in target_factor_outputs:
+                if not self._skip_softmax:
+                    tf_logits = pt.log_softmax(tf_logits, dim=-1)
+                tf_scores = -tf_logits
+                # shape: (batch, num_classes)
+                # target factors are greedily chosen, and score and index are collected via torch.max.
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/beam_search_pt.py:90]: done

File: sockeye/beam_search_pt.py [hunk 3]
```diff
@@ -77,13 +77,23 @@ def decode_step(self,
...
         if target_factor_outputs:
-            # target factors are greedily 'decoded'.
-            factor_predictions = [tfo.argmax(dim=1).unsqueeze(1).int() for tfo in target_factor_outputs]
-            target_factors = factor_predictions[0] if len(factor_predictions) == 1 else pt.cat(factor_predictions, 1)
+            predictions = []  # type: List[pt.Tensor]
+            for tf_logits in target_factor_outputs:
+                if not self._skip_softmax:
+                    tf_logits = pt.log_softmax(tf_logits, dim=-1)
+                tf_scores = -tf_logits
+                # shape: (batch, num_classes)
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/beam_search_pt.py:89]: no, just a leftover. Will remove.
